### ISSUE TITLE
Taking `requires_grad` propagation seriously in Thunder

### ISSUE THREAD

**Issue by IvanYashchuk** [Lightning-AI/lightning-thunder#1768]
## 🚀 Feature

Add correct `requires_grad` attribute propagation in Thunder with OpInfo-based tests.

Currently, Thunder queries the `requires_grad` attribute only for compiled function inputs (nn.Module parameters, buffers, and args) and uses this attribute to cache compiled artifacts and mask the output of the compiled backward function.

We could also faithfully model the propagation of this attribute, as in PyTorch Eager, or consistently, according to the rules that we need to develop. However, the exact benefits are unclear and need to be collected to consider whether it's worth the effort.

If we decide to do the `requires_grad` propagation we would be able to faithfully mark some outputs of the compiled function with `requires_grad=False` by disconnecting them from the Autograd graph (https://github.com/Lightning-AI/lightning-thunder/issues/1733). The performance benefits of doing so remain to be evaluated.

**Comment by t-vi** [comment]
So my thinking is that globally:

- we have requires_grad on the inputs (easy!),
- we need to propagate automatically during the autodiff pass, so that is where we determine which outputs need grad,
- we need to keep it correct on outputs in further transformations.

In the autodiff pass:

- bsyms things that require grads on their outputs can things requiring grad as output by default (limited to floating point tensors),
- we only want to materialize `put_grad` for tensors that needing gradients and then "dce" things,
- the things that require grad are the `get_grad`s we have left, 
- there likely needs to be an escape hatch, but probably only for the grad_transform (joint forward / backward) style for things that have a requires_grad pattern that cannot be automatically inferred, so those need to set requires_grad as needed.

Probably one simple interesting case is `a * b` with various combinations of `requires_grad` on `a` and `b`.
